ci/bin/format.dart

```diff
@@ -0,0 +1,795 @@
+    }
+    for (String file in files) {
+      if (file.trim().isEmpty) {
+        continue;
+      }
+      formatJobs.add(WorkerJob(<String>['java', '-jar', googleJavaFormatJar.path, file.trim()]));
+    }
+    final ProcessPool formatPool =
+        ProcessPool(processRunner: _processRunner, printReport: namedReport('Java format'));
+    final Stream<WorkerJob> completedClangFormats = formatPool.startWorkers(formatJobs);
+    final List<WorkerJob> diffJobs = <WorkerJob>[];
+    await for (final WorkerJob completedJob in completedClangFormats) {
+      if (completedJob.result != null && completedJob.result.exitCode == 0) {
+        diffJobs.add(WorkerJob(<String>['diff', '-u', completedJob.command.last, '-'],
+            stdinRaw: codeUnitsAsStream(completedJob.result.stdoutRaw), failOk: true));
+      }
+    }
+    final ProcessPool diffPool =
+        ProcessPool(processRunner: _processRunner, printReport: namedReport('diff'));
+    final List<WorkerJob> completedDiffs = await diffPool.runToCompletion(diffJobs);
+    final Iterable<WorkerJob> failed = completedDiffs.where((WorkerJob job) {
+      return job.result.exitCode != 0;
+    });
+    reportDone();
+    if (failed.isNotEmpty) {
+      final bool plural = failed.length > 1;
+      if (fixing) {
+        error('Fixing ${failed.length} Java file${plural ? 's' : ''}'
+            ' which ${plural ? 'were' : 'was'} formatted incorrectly.');
+      } else {
+        error('Found ${failed.length} Java file${plural ? 's' : ''}'
+            ' which ${plural ? 'were' : 'was'} formatted incorrectly.');
+        for (final WorkerJob job in failed) {
+          stdout.write(job.result.stdout);
+        }
+      }
+    } else {
+      message('Completed checking ${diffJobs.length} Java files with no formatting problems.');
+    }
+    return failed.map<String>((WorkerJob job) {
+      return job.result.stdout;
+    }).toList();
+  }
+}
+
+/// Checks the format of any BUILD.gn files using check_gn_format.py.
+class GnFormatChecker extends FormatChecker {
+  GnFormatChecker({
+    ProcessManager /*?*/ processManager,
+    @required String baseGitRef,
+    @required Directory repoDir,
+    @required Directory srcDir,
+    bool allFiles = false,
+    MessageCallback messageCallback,
+  }) : super(
+          processManager: processManager,
+          baseGitRef: baseGitRef,
+          repoDir: repoDir,
+          srcDir: srcDir,
+          allFiles: allFiles,
+          messageCallback: messageCallback,
+        );
+
+  @override
+  Future<bool> checkFormatting() async {
+    message('Checking GN formatting...');
+    return (await _runGnCheck(fixing: false)) == 0;
+  }
+
+  @override
+  Future<bool> fixFormatting() async {
+    message('Fixing GN formatting...');
+    await _runGnCheck(fixing: true);
+    // The GN script doesn't fail when fixing errors.
+    return true;
+  }
+
+  Future<int> _runGnCheck({@required bool fixing}) async {
+    final List<String> cmd = <String>[
+      'python',
```

Comment:
doesn't have to be this patch, but it's a little strange to me to have this utility exist in python when everything else in here is Dart or some other native thing maintained by others. I suspect we really want this to use the same file inding logic in here rather than being duplicated.

The call to `gn format` just needs a `-w` for fixing. Up to you whether you want to include that in this patch or just keep it as a rewrite o the bash script for now, but we should probably at least add a todo for it.